helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -165,6 +166,27 @@ public void recordPathStat(Stat stat, OptionalLong notificationTime) {
     }
   }
 
+  private abstract class SessionAwareCallable<T> implements Callable<T> {
+    protected volatile String sessionId;
+    protected volatile ZooKeeper zookeeper;
```

Comment:
When this Callable gets instantiated with a constructor that takes in both sessionId and zookeeper, please make sure you validate them as well. We could even do the check for whether sessionId given and ZooKeeper's sessionId match in the constructor and fail early?

Reply:
Actually, let's be careful here. We should only fail if mode is ephemeral.